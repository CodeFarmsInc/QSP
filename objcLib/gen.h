#ifndef ZZ_CODEGEN_INCLUDE
#define ZZ_CODEGEN_INCLUDE

#include "lib/general.h"

class Generator;
class Organization;
class Participant;
class OrgType;
class PartType;
class Iterator;
class FixedType;
class ClassLink;
class ApplClass;

#include "tmp/orgs.h"
#include "tmp/orgs1.h"
#include "tmp/parts.h"
#include "tmp/parts1.h"
#include "tmp/parts2.h"
#include "tmp/orgName.h"
#include "tmp/partName.h"
#include "tmp/types.h"
#include "tmp/types1.h"
#include "tmp/partTypes.h"
#include "tmp/partTypes1.h"
#include "tmp/partTypes2.h"
#include "tmp/inherits.h"
#include "tmp/iters.h"
#include "tmp/iters1.h"
#include "tmp/orgTypeName.h"
#include "tmp/fileName.h"
#include "tmp/partTypeName.h"
#include "tmp/iterName.h"
#include "tmp/inhName.h"
#include "tmp/fixedName.h"
#include "tmp/inhMatch.h"
#include "tmp/inhFixed.h"
#include "tmp/fixedTypes.h"
#include "tmp/fixedTypes1.h"
#include "tmp/myType.h"
#include "tmp/myPartType.h"
#include "tmp/toBase.h"
#include "tmp/classes.h"
#include "tmp/classes1.h"
#include "tmp/dependsOn.h"
#include "tmp/dependsOn1.h"
#include "tmp/usedBy.h"
#include "tmp/usedBy1.h"
#include "tmp/usedBy2.h"
#include "tmp/cName.h"

// -----------------------------------------------
typedef class orgs_LinkedList2 orgs;
typedef class parts_Aggregate2 parts;
typedef class orgName_Name orgName;
typedef class partName_Name partName;
typedef class types_LinkedList2 types;
typedef class partTypes_Aggregate2 partTypes;
typedef class inherits_SingleLink inherits;
typedef class iters_LinkedList1 iters;
typedef class orgTypeName_Name orgTypeName;
typedef class fileName_Name fileName;
typedef class partTypeName_Name partTypeName;
typedef class iterName_Name iterName;
typedef class inhName_Name inhName;
typedef class fixedName_Name fixedName;
typedef class inhMatch_SingleLink inhMatch;
typedef class inhFixed_SingleLink inhFixed;
typedef class fixedTypes_LinkedList1 fixedTypes;
typedef class myType_SingleLink myType;
typedef class myPartType_SingleLink myPartType;
typedef class toBase_SingleLink toBase;
typedef class classes_Hash classes;
typedef class dependsOn_LinkedList1 dependsOn;
typedef class usedBy_Aggregate2 usedBy;
typedef class cName_Name cName;
// -----------------------------------------------


class ZZ_Generator {
public:
    orgs_LinkedList2Parent ZZorgs;
    types_LinkedList2Parent ZZtypes;
    classes_HashHolder ZZclasses;
};

class ZZ_Organization {
public:
    orgs_LinkedList2Child ZZorgs;
    parts_Aggregate2Parent ZZparts;
    orgName_NameParent ZZorgName;
    myType_SingleLinkParent ZZmyType;
};

class ZZ_Participant {
public:
    parts_Aggregate2Child ZZparts;
    partName_NameParent ZZpartName;
    myPartType_SingleLinkParent ZZmyPartType;
    usedBy_Aggregate2Child ZZusedBy;
};

class ZZ_OrgType {
public:
    types_LinkedList2Child ZZtypes;
    partTypes_Aggregate2Parent ZZpartTypes;
    inherits_SingleLinkParent ZZinherits;
    iters_LinkedList1Parent ZZiters;
    orgTypeName_NameParent ZZorgTypeName;
    fileName_NameParent ZZfileName;
    fixedTypes_LinkedList1Parent ZZfixedTypes;
};

class ZZ_PartType {
public:
    partTypes_Aggregate2Child ZZpartTypes;
    partTypeName_NameParent ZZpartTypeName;
    inhName_NameParent ZZinhName;
    inhMatch_SingleLinkParent ZZinhMatch;
};

class ZZ_Iterator {
public:
    iters_LinkedList1Child ZZiters;
    iterName_NameParent ZZiterName;
};

class ZZ_FixedType {
public:
    fixedName_NameParent ZZfixedName;
    inhFixed_SingleLinkParent ZZinhFixed;
    fixedTypes_LinkedList1Child ZZfixedTypes;
};

class ZZ_ClassLink {
public:
    toBase_SingleLinkParent ZZtoBase;
    dependsOn_LinkedList1Child ZZdependsOn;
};

class ZZ_ApplClass {
public:
    classes_HashElement ZZclasses;
    dependsOn_LinkedList1Parent ZZdependsOn;
    usedBy_Aggregate2Parent ZZusedBy;
    cName_NameParent ZZcName;
};

typedef orgs_LinkedList2Iterator orgs_Iterator;
typedef parts_Aggregate2Iterator parts_Iterator;
typedef types_LinkedList2Iterator types_Iterator;
typedef partTypes_Aggregate2Iterator partTypes_Iterator;
typedef iters_LinkedList1Iterator iters_Iterator;
typedef fixedTypes_LinkedList1Iterator fixedTypes_Iterator;
typedef classes_HashIterator classes_Iterator;
typedef dependsOn_LinkedList1Iterator dependsOn_Iterator;
typedef usedBy_Aggregate2Iterator usedBy_Iterator;

#endif // ZZ_CODEGEN_INCLUDE
