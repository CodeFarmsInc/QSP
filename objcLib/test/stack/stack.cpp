
// ------------------------------------------------------
// This is the stack part from the DOL test9a coded in the "static style.

// The test problem:
// A set of employee records is loaded into a ring (eRing).
// and from there into the stack.
// Two sets of tests are made: one for the stack of full employee
// records(eStack), the other for the array of pointers (pStack).
// The arrays start with a smaller size than actually
// needed, so that the re-allocation mechanism is tested.

// Input format: employeeName id salary
// ------------------------------------------------------

#include <stdio.h>
#include <string.h>
#include "gen.h"

class Header {
public:
    ZZ_Header ZZds;
    void freeArr();
    void cpEmpl(Employee *p1,Employee *p2);
};
class Employee {
public:
    ZZ_Employee ZZds;
    int emplNum; // employee number or id
    int salary;
    int heapIndex;
};

#define BSIZE 80

main(int argc,char **argv) {
    FILE *file1,*file2;
    Header *h;
    Employee *ep,*sp,**pa,*oa,*p1,**p2,*ee;
    char buff[BSIZE],name[BSIZE],*cp,*cc;
    int n;
    eRing_Iterator eIter;

    if(argc<=1)file1=stdin;  else file1=fopen(argv[1],"r");
    if(argc<=2)file2=stdout; else file2=fopen(argv[2],"w");

    h=new Header;    /* main holder of the arrays */
    ee=new Employee; /* temporary working object */
    sp=(Employee *)NULL;
    n=0;
    while(fgets(buff,BSIZE,file1)){
        ep=new Employee;
        sscanf(buff,"%s %d %d\n",name,&(ep->emplNum),&(ep->salary));
        cp=new char[strlen(name)+1]; strcpy(cp,name);
        eName::add(ep,cp);
        sp=eRing::add(sp,ep);
        n++;
    }
    rStart::add(h,sp);
    fclose(file1);

    fprintf(file2,"\n using the array as a stack\n");
    pa=pStack::form(h,4,-3);
    oa=eStack::form(h,3,10);

    eIter.start(sp);
    for(ep= ++eIter; ep; ep= ++ eIter){
        p2= &ep;
        pStack::push(h,p2);
        h->cpEmpl(ep,ee);
        eStack::push(h,ee);
        cp=eName::get(ep);
        fprintf(file2,"push: %s emplNum=%d salary=%d\n",
                                                 cp,ep->emplNum,ep->salary);
    }
    for(p1=sp;p1;){ /* start with any non-null pointer from loading eRing */
        p2=pStack::pop(h);
        if(!p2)break;
        cc=eName::get(*p2);
        p1=eStack::pop(h);
        if(!p1)break;
        cp=eName::get(p1);
        fprintf(file2,"pop: %s emplNum=%d salary=%d",
                                                 cp,p1->emplNum,p1->salary);
        if(!strcmp(cc,cp))fprintf(file2,"\n");
        else fprintf(file2," different names: %s %s\n",cc,cp);
    }

    h->freeArr();
    return(0);
}
/* --------------------------------------------------------------------- */
/* disconnect links on array oa, and free both arrays */
void Header::freeArr()
{
    int i,sz;
    Employee *oa;

    sz=eStack::size(this);
    oa=eStack::ind(this,0);
    for(i=0;i<sz;i++){
        eName::del(&(oa[i]));
    }
    pStack::free(this);
    eStack::free(this);
    return;
}
/* --------------------------------------------------------------------- */
/* Copy one Employee record onto another, but avoid the organization pointers.
 * Note that: (*p2)=(*p1); would be wrong, it would mess up the pointers */
void Header::cpEmpl(Employee *p1,Employee *p2)
{
    char *cp;

    eName::del(p2);
    cp=eName::get(p1);
    eName::add(p2,cp); /* using the same name as string */
    p2->emplNum=p1->emplNum;
    p2->salary=p1->salary;
    return;
}

#include "gen.cpp"
