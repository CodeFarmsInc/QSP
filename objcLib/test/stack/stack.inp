J.Brown 4367 2105
A.Green 1876 3500
I.Black 7003 2795
S.White 3523 2400
P.Grey  1289 4140
F.Gold  7990 3905
L.Dotty 7543 3230
J.Shade 1987 1900
E.Light 3387 2045
W.Water 1100 2400
T.Seal  2901 2400
