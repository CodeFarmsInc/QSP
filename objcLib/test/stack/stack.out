
 using the array as a stack
push: T.Seal emplNum=2901 salary=2400
push: W.Water emplNum=1100 salary=2400
push: E.Light emplNum=3387 salary=2045
push: J.Shade emplNum=1987 salary=1900
push: L.Dotty emplNum=7543 salary=3230
push: F.Gold emplNum=7990 salary=3905
push: P.Grey emplNum=1289 salary=4140
push: S.White emplNum=3523 salary=2400
push: I.Black emplNum=7003 salary=2795
push: A.Green emplNum=1876 salary=3500
push: J.Brown emplNum=4367 salary=2105
pop: J.Brown emplNum=4367 salary=2105
pop: A.Green emplNum=1876 salary=3500
pop: I.Black emplNum=7003 salary=2795
pop: S.White emplNum=3523 salary=2400
pop: P.Grey emplNum=1289 salary=4140
pop: F.Gold emplNum=7990 salary=3905
pop: L.Dotty emplNum=7543 salary=3230
pop: J.Shade emplNum=1987 salary=1900
pop: E.Light emplNum=3387 salary=2045
pop: W.Water emplNum=1100 salary=2400
pop: T.Seal emplNum=2901 salary=2400
